pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/HashJoinOperator.java

```diff
@@ -43,50 +49,94 @@
 
 /**
  * This basic {@code BroadcastJoinOperator} implement a basic broadcast join algorithm.
+ * This algorithm assumes that the broadcast table has to fit in memory since we are not supporting any spilling.
  *
+ * For left join, inner join, right join and full join,
  * <p>It takes the right table as the broadcast side and materialize a hash table. Then for each of the left table row,
  * it looks up for the corresponding row(s) from the hash table and create a joint row.
  *
  * <p>For each of the data block received from the left table, it will generate a joint data block.
- *
- * We currently support left join, inner join and semi join.
+ * We currently support left join, inner join, right join and full join.
  * The output is in the format of [left_row, right_row]
  */
 // TODO: Move inequi out of hashjoin. (https://github.com/apache/pinot/issues/9728)
 public class HashJoinOperator extends BaseOperator<TransferableBlock> {
   private static final String EXPLAIN_NAME = "HASH_JOIN";
-  private static final Set<JoinRelType> SUPPORTED_JOIN_TYPES = ImmutableSet.of(JoinRelType.INNER, JoinRelType.LEFT);
-  private final HashMap<Key, List<Object[]>> _broadcastHashTable;
+  private static final Set<JoinRelType> SUPPORTED_JOIN_TYPES =
+      ImmutableSet.of(JoinRelType.INNER, JoinRelType.LEFT, JoinRelType.RIGHT, JoinRelType.FULL);
+
+  private final HashMap<Key, List<Object[]>> _broadcastRightTable;
+
+  // Used to track matched right rows.
+  // Only used for right join and full join to output non-matched right rows.
+  private final HashMap<Key, HashSet<Integer>> _matchedRightRows;
+
+  // Used to track unmatched left rows.
+  // Only used for left join and full join to output non-matched left rows.
+  private final HashMap<Key, List<Object[]>> _unmatchedLeftRows;
+
   private final Operator<TransferableBlock> _leftTableOperator;
   private final Operator<TransferableBlock> _rightTableOperator;
   private final JoinRelType _joinType;
   private final DataSchema _resultSchema;
+  private final int _leftRowSize;
   private final int _resultRowSize;
   private final List<TransformOperand> _joinClauseEvaluators;
   private boolean _isHashTableBuilt;
+
+  // Used by non-inner join.
+  // Needed to indicate we have finished processing all results after returning last block.
+  // TODO: Remove this special handling by fixing data block EOS abstraction or operator's invariant.
+  private boolean _isTerminated;
   private TransferableBlock _upstreamErrorBlock;
   private KeySelector<Object[], Object[]> _leftKeySelector;
   private KeySelector<Object[], Object[]> _rightKeySelector;
 
+  private boolean needUnmatchedRightRows() {
+    return _joinType == JoinRelType.RIGHT || _joinType == JoinRelType.FULL;
+  }
+
+  private boolean needUnmatchedLeftRows() {
+    return _joinType == JoinRelType.LEFT || _joinType == JoinRelType.FULL;
+  }
+
+  private boolean needUnmatchedRows() {
+    return _joinType != JoinRelType.INNER;
+  }
```

Comment:
nit: private util methods are usually put after the public ones in pinot. so please move these to the end of file

Reply:
done.